src/gt4py/caching.py

```diff
@@ -0,0 +1,412 @@
+import pathlib
+import pickle
+import sys
+import types
+from typing import Any, Dict, Optional, List, TYPE_CHECKING
+
+import gt4py
+import gt4py.definitions
+
+
+if TYPE_CHECKING:
+    from gt4py.stencil_builder import StencilBuilder
+
+
+class CachingStrategy(abc.ABC):
+    name: str
+
+    def __init__(self, builder: "StencilBuilder"):
+        self.builder = builder
+
+    @property
+    @abc.abstractmethod
+    def root_path(self) -> pathlib.Path:
+        """
+        Get the root file path for stencil source, build and cache files.
+
+        Returns
+        -------
+        type: Pathlike
+
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def backend_root_path(self) -> pathlib.Path:
+        """
+        Get the caching base path for the backend being used.
+
+        Returns
+        -------
+        type: Pathlike
+
+        """
+        raise NotImplementedError
+
+    @property
+    @abc.abstractmethod
+    def cache_info_path(self) -> Optional[pathlib.Path]:
+        """
+        Get the cache info file path for the current build process.
+
+        Returns
+        -------
+        type: Pathlike or None
+
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def generate_cache_info(self) -> Dict[str, Any]:
+        """
+        Generate the cache info dict, adding in additional info passed.
+        Backends can overload :py:meth:`gt4py.backend.base.Backend.extra_cache_info` to store extra
+        info.
+
+        Returns
+        -------
+        type: Dict[str, Any]
+            The generated cache info dictionary
+
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def update_cache_info(self) -> None:
+        """
+        Regenerate cache info and update the cache info file.
+
+        Parameters
+        ----------
+        extra_cache_info: Dic[str, Any]
+            Additional info to be written
+
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def validate_cache_info(self, cache_info: Dict[str, Any], *, validate_hash: bool,) -> bool:
+        """
+        Validate the passed cache info by comparing it to the stored data.
+
+        A backend can cause additional keys to be validated by overloading
+        :py:meth:`gt4py.backend.base.Backend.extra_cache_validate`.  These keys must also be
+        present in it's overload of :py:meth:`gt4py.backend.base.Backend.extra_cache_info`.
+
+        Parameters
+        ----------
+        cache_info: Dict[str, Any]
+            The cache info to be compared to file
+
+        validate_hash: bool
+            Should the stencil fingerprint and source code hash be validated or not
+
+        Returns
+        -------
+        type: bool
+            Is the given cache info consistent with what's stored on file
+        """
+        raise NotImplementedError
+
+    @abc.abstractmethod
+    def check_cache_info(self, *, validate_hash: bool) -> bool:
```

Comment:
I'd prefer a name that states more clearly what the result of the check is, e.g. `is_cache_info_available` or `is_cache_info_consistent`. (I read a test first and it wasn't clear to me what the check does.)

Reply:
Renamed for more clarity